### PR TITLE
CLN: _wrap_applied_output

### PR DESCRIPTION
**Pull request by rhshadrach**
The majority of this change is the result of three operations:

1. Move code that returns up top, to avoid the heavily nested structure.
2. Move variable creation as close as possible to where they are used. Previously, in certain cases computations were being done and then going unused.
3. Combine duplicated code.

There are two sections that I was able to entirely remove:

````
ping = self.grouper.groupings[0]
if len(keys) == ping.ngroups:
    key_index = ping.group_index
    key_index.name = key_names[0]

    key_lookup = Index(keys)
    indexer = key_lookup.get_indexer(key_index)

    # reorder the values
    values = [values[i] for i in indexer]

    # update due to the potential reorder
    first_not_none = next(com.not_none(*values), None)
````

and

````
# GH5788 instead of stacking; concat gets the
# dtypes correct
from pandas.core.reshape.concat import concat

result = concat(
    values,
    keys=key_index,
    names=key_index.names,
    axis=self.axis,
).unstack()
result.columns = index
````

For the single test that I touched, it was checking that the categorical dtype of an index was being dropped after a groupby. I don't believe that is the correct behavior - that the categorical dtype should remain. I checked groupby/categorical bugs and didn't find any issues this closes.

### CODE REVIEW

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -1213,171 +1213,112 @@ def _wrap_applied_output(self, keys, values, not_indexed_same=False):
...
 
-                    # reorder the values
-                    values = [values[i] for i in indexer]
+        key_index = self.grouper.result_index if self.as_index else None
+        v = values[0]
 
-                    # update due to the potential reorder
-                    first_not_none = next(com.not_none(*values), None)
-                else:
+        if not isinstance(v, (np.ndarray, Index, Series)):
```
Comment (WillAyd) [inline pandas/core/groupby/generic.py:1243]: Can this just be an else statement? Or are there more types we handle than these + NDFrame?

Comment (rhshadrach) [reply pandas/core/groupby/generic.py:1243]: I think you're suggesting changing this to read 

    if isinstance(v, (np.ndarray, Index, Series)):
        ...
    else:
        ...

Doing so means the rest of this method is nested within the first `if` clause, which is what I was trying to avoid. Having this nesting separates the else from the if, and reading that deeply nested structure is

Comment (rhshadrach) [reply pandas/core/groupby/generic.py:1243]: I think you're suggesting something like:

````
if isinstance(v, (np.ndarray, Index, Series)):
    ...
else:
    ...
````

The reason I have opted not to do this is that the if-block is exceedingly long, whereas the else-block is quite short. Doing it this way would result in a more nested rather than flat structure.

Comment (rhshadrach) [reply pandas/core/groupby/generic.py:1243]: Ack - sorry, I see what you're saying now. Ignore my previous response, will investigate.

File: pandas/core/indexes/api.py [hunk 2]
```diff
@@ -298,15 +298,16 @@ def all_indexes_same(indexes):
...
+    indexes : iterable of Index objects
 
     Returns
     -------
     bool
         True if all indexes contain the same elements, False otherwise.
     """
-    first = indexes[0]
-    for index in indexes[1:]:
+    itr = iter(indexes)
```
Comment (WillAyd) [inline pandas/core/indexes/api.py:308]: Is this change related to the PR?

File: pandas/tests/groupby/test_apply.py [hunk 3]
```diff
@@ -868,13 +868,15 @@ def test_apply_multi_level_name(category):
     b = [1, 2] * 5
     if category:
         b = pd.Categorical(b, categories=[1, 2, 3])
+        expected_index = pd.CategoricalIndex([1, 2], categories=[1, 2, 3], name="B")
```
Comment (WillAyd) [inline pandas/tests/groupby/test_apply.py:871]: Is there an open issue for this?

**Review by WillAyd — COMMENTED** [review]
Thanks for refactoring this - these code paths as I'm sure you've noticed are thorny, so cleanups here are much appreciated

File: pandas/core/indexes/api.py (hunk 2 above)
Comment (rhshadrach) [reply pandas/core/indexes/api.py:308]: Yes - it allows for this line:

    all_indexed_same = all_indexes_same((x.index for x in values))

Previously a full list was created and then passed to `all_indexes_same`. By using a generator, we allow the possibility to short-circuit if say the first to indices are different.

File: pandas/tests/groupby/test_apply.py (hunk 3 above)
Comment (rhshadrach) [reply pandas/tests/groupby/test_apply.py:871]: I don't believe so. There is only one issue tagged with categorical, groupby, and apply which is not relevant. I also took a  look through those tagged as categorical and groupby and didn't see anything either.

**Comment by rhshadrach** [comment]
Codecov is finding a lot in this method that isn't hit. I'm going to see what tests can be added or branches that can be removed.

File: pandas/tests/groupby/test_apply.py (hunk 3 above)
Comment (WillAyd) [reply pandas/tests/groupby/test_apply.py:864]: Can you add a whatsnew for this? Something along the lines of `groupby apply will now maintain a CategoricalIndex` (assuming that is now the case)

**Review by WillAyd — COMMENTED** [review]
Cool I think this is a nice refactor @jreback

**Comment by jbrockmendel** [comment]
> The majority of this change is the result of three operations:

To the extent that you can split this into independent pieces, it will be easier to review.
